File: examples/dod/did_rpc_loopback.cpp
#include <chrono>
#include <iostream>
#include <optional>

#include <authbox.hpp>
#include <pki/builder.hpp>
#include <pki/key_utils.hpp>

namespace {

    struct LoopbackRemote {
        authbox::did::rpc::Service *service{nullptr};

        dp::Res<netpipe::Message> call(dp::u32 method_id, const netpipe::Message &request, dp::u32) {
            if (!service) {
                return dp::result::err(dp::Error::not_found("service not configured"));
            }
            return service->dispatch(method_id, request);
        }
    };

} // namespace

int main() {
    const std::string did_uri = "did:web:example.com";

    auto keypair = authbox::pik::generate_ed25519_keypair();
    authbox::pik::CertificateBuilder builder;
    const auto now = std::chrono::system_clock::now();

    authbox::pik::SubjectAltNameExtension::GeneralName did_name{};
    did_name.type = authbox::pik::SubjectAltNameExtension::GeneralNameType::URI;
    did_name.value = did_uri;

    auto cert = builder.set_subject_from_string("CN=example.com,O=Example")
                    .set_issuer_from_string("CN=example.com,O=Example")
                    .set_validity(now - std::chrono::hours(1), now + std::chrono::hours(24))
                    .set_subject_public_key_ed25519(keypair.public_key)
                    .set_basic_constraints(false, std::nullopt)
                    .set_subject_alt_name({did_name})
                    .build_ed25519(keypair, true);
    if (!cert.success) {
        std::cerr << "Failed to create certificate: " << cert.error << "\n";
        return 1;
    }

    auto generated = authbox::did::generate_did_web_document("example.com", cert.value);
    if (generated.is_err()) {
        std::cerr << "Failed to generate did document: " << generated.error().c_str() << "\n";
        return 1;
    }

    authbox::did::Resolver resolver([&generated](std::string_view url) -> authbox::did::DidResult<std::string> {
        if (url != "https://example.com/.well-known/did.json") {
            return authbox::did::DidResult<std::string>::err(authbox::did::to_dp_string("unexpected URL"));
        }
        return authbox::did::DidResult<std::string>::ok(
            std::string(generated.value().did_document_json.data(), generated.value().did_document_json.size()));
    });

    authbox::did::rpc::Service service(resolver);
    LoopbackRemote remote{&service};
    authbox::did::rpc::Client<LoopbackRemote> client(remote);

    auto resolved = client.resolve(did_uri);
    if (resolved.is_err()) {
        std::cerr << "RPC resolve failed: " << resolved.error().c_str() << "\n";
        return 1;
    }

    auto verified = client.verify_binding(
        did_uri,
        std::string_view(generated.value().did_document_json.data(), generated.value().did_document_json.size()),
        cert.value.to_pem());
    if (verified.is_err() || !verified.value()) {
        std::cerr << "RPC verify failed: " << (verified.is_err() ? verified.error().c_str() : "invalid") << "\n";
        return 1;
    }

    std::cout << "RPC resolve source: " << resolved.value().source_url.c_str() << "\n";
    std::cout << "RPC verify: OK\n";
    return 0;
}

File: examples/did/did_rpc_loopback.rs
use authbox::did::{
    Resolver, generate_did_web_document_from_certificate,
    rpc::{Client as DidRpcClient, LoopbackRemote, Service},
};
use authbox::pki::{
    CertificateBuilder, DerTime, GeneralName, GeneralNameType, generate_ed25519_keypair,
};

fn example_time(year: i32) -> DerTime {
    DerTime {
        year,
        month: 1,
        day: 1,
        hour: 0,
        minute: 0,
        second: 0,
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let did_uri = "did:web:example.com";
    let keypair = generate_ed25519_keypair()?;
    let did_name = GeneralName {
        type_: GeneralNameType::Uri,
        value: did_uri.as_bytes().to_vec(),
    };

    let cert = CertificateBuilder::new()
        .set_subject_from_string("CN=example.com,O=Example")?
        .set_issuer_from_string("CN=example.com,O=Example")?
        .set_validity(example_time(2020), example_time(2035))
        .set_subject_public_key_ed25519(keypair.public_key.clone())
        .set_basic_constraints(false, None)?
        .set_subject_alt_name(&[did_name])?
        .build_ed25519_with_self_signed(&keypair, true)?;

    let generated = generate_did_web_document_from_certificate("example.com", &cert)?;
    let expected_doc = generated.did_document_json.clone();
    let resolver = Resolver::with_fetcher(move |url| {
        if url == "https://example.com/.well-known/did.json" {
            Ok(expected_doc.clone())
        } else {
            Err(authbox::did::error::network_error("unexpected URL"))
        }
    });

    let service = Service::new(resolver);
    let remote = LoopbackRemote::new(service);
    let client = DidRpcClient::new(remote);

    let resolved = client.resolve(did_uri)?;
    let verified = client.verify_binding(did_uri, &generated.did_document_json, &cert.to_pem())?;

    if !verified {
        return Err("RPC verify returned invalid binding".into());
    }

    println!("RPC resolve source: {}", resolved.source_url);
    println!("RPC verify: OK");

    Ok(())
}
